query-engine/core/src/query_graph_builder/write/delete.rs

```diff
@@ -104,3 +154,90 @@ pub fn delete_many_records(
 
     Ok(())
 }
+
+/// An atomic delete is a delete performed in a single operation. It uses `DELETE ... RETURNING` or
+/// similar statement.
+/// We only perform such delete when:
+/// 1. Connector supports such operations
+/// 2. The selection set contains no relation
+fn can_use_atomic_delete(query_schema: &QuerySchema, model: &Model, field: &ParsedField<'_>) -> bool {
+    if !query_schema.has_capability(ConnectorCapability::DeleteReturning) {
+        return false;
+    }
+
+    if field.has_nested_selection() {
+        return false;
+    }
+
+    let relations_selected = field
+        .nested_fields
+        .as_ref()
+        .unwrap()
+        .fields
+        .iter()
+        .any(|pair| pair.schema_field.maps_to_relation());
+    if relations_selected && !query_schema.has_capability(ConnectorCapability::SupportsFiltersOnRelationsWithoutJoins) {
+        return false;
+    }
+
+    // `ReferentialAction::Restrict` action requires us to check if any connected records exist
+    // before performing delete. To do that, we need ids of records to be deleted. Atomic delete
+    // would provide us with ids, but it would also delete the record immediately, so we won't be
+    // able to check anything. So in this case we resort to a plan with a separate read node,
+    // allowing us to perform the check before deleting anything.
+    // TODO laplab: expand comment.
+    if !query_schema.relation_mode().uses_foreign_keys() && has_restrict_on_delete(model) {
+        return false;
+    }
+
+    true
+}
+
+fn has_restrict_on_delete(model: &Model) -> bool {
+    let internal_model = &model.dm;
+    let relation_fields = internal_model.fields_pointing_to_model(model);
+    for relation_field in relation_fields {
+        match relation_field.relation().on_delete() {
+            ReferentialAction::NoAction | ReferentialAction::Restrict => return true,
+            ReferentialAction::SetNull => {
+                let overlapping_fields = collect_overlapping_relation_fields(relation_field.model(), &relation_field);
+                if has_restrict_on_update(overlapping_fields.into_iter()) {
+                    return true;
+                }
+            }
+            ReferentialAction::Cascade => {
+                if has_restrict_on_delete(&relation_field.model()) {
+                    return true;
+                }
+            }
```

Comment:
Could you elaborate on the rationale behind this?

Reply:
Confirmed it, there is no need for these checks with the transaction present. I removed this code.